transport/cluster.go

```diff
@@ -0,0 +1,295 @@
+package transport
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strconv"
+	"time"
+
+	"scylla-go-driver/frame"
+	. "scylla-go-driver/frame/response"
+
+	"go.uber.org/atomic"
+)
+
+type PeerMap = map[string]*Node
+
+type eventHandler = responseHandler
+
+type refreshHandler chan struct{}
+
+type Cluster struct {
+	peers         atomic.Value // PeerMap
+	control       *Conn
+	cfg           ConnConfig
+	events        eventHandler
+	handledEvents []frame.EventType
+	knownHosts    []string
+	refresher     refreshHandler
+
+	// Yet to be used.
+	// ring        map[Token]*Node
+	// dataCenters map[string]*DataCenter
+}
+
+// type dataCenter struct {
+// 	nodes   []Node
+// 	rackCnt uint // Type?
+// }
+
+// NewCluster also creates control connection and starts handling events and refreshing topology.
+func NewCluster(cfg ConnConfig, e []frame.EventType, hosts ...string) (*Cluster, error) {
+	if len(hosts) < 1 {
+		return nil, fmt.Errorf("at least one host is required to create cluster")
+	}
+
+	p := atomic.Value{}
+	p.Store(PeerMap{})
+	k := make([]string, len(hosts))
+	copy(k, hosts)
+
+	c := &Cluster{
+		peers:         p,
+		cfg:           cfg,
+		handledEvents: e,
+		knownHosts:    k,
+	}
+
+	if err := c.NewControl(); err != nil {
+		return nil, fmt.Errorf("creating cluster: %w", err)
+	}
+	if err := c.refreshTopology(); err != nil {
+		return nil, fmt.Errorf("creating cluster: %w", err)
+	}
+
+	go c.loop()
+	return c, nil
+}
+
+func (c *Cluster) NewControl() error {
+	log.Printf("opening control connection")
+	for _, v := range c.knownHosts {
+		control, err := OpenConn(v, nil, c.cfg)
+		if err == nil {
+			if events, err := control.registerEvents(c.handledEvents); err == nil {
+				c.control = control
+				c.events = events
+
+				go c.handleEvents()
+				return nil
+			}
+		}
+		if control != nil {
+			control.Close()
+		}
+	}
+	return fmt.Errorf("couldn't open control connection to any known host")
+}
+
+// refreshTopology creates new PeerMap filled with the result of both localQuery and peerQuery.
+// The old map is replaced with the new one atomically to prevent dirty reads.
+func (c *Cluster) refreshTopology() error {
+	log.Printf("refreshing topology")
+	rows, err := c.getAllNodesInfo()
+	if err != nil {
+		return fmt.Errorf("refreh topology: %w", err)
+	}
+	// If node is present in both maps we can update already created node
+	// instead of creating new one from the scratch.
+	old := c.GetPeers()
+	m := PeerMap{}
+	for _, v := range rows {
+		addr := net.IP(v[nodeAddr]).String()
+
+		if node, ok := old[addr]; ok {
+			node.Tokens = v[nodeTokens]
+			m[addr] = node
+		} else {
+			n := &Node{
+				Addr:       addr,
+				Datacenter: string(v[nodeDC]),
+				Rack:       string(v[nodeRack]),
+				Tokens:     v[nodeTokens],
+			}
+			if pool, err := NewConnPool(appendDefaultPort(addr), c.cfg); err != nil {
```

Comment:
Please do not append it here.
It should be appended in NewConnPool!

Reply:
There is some discussion related to that [in this PR comment](https://github.com/mmatczuk/scylla-go-driver/pull/148#pullrequestreview-906906584).